### ISSUE TITLE
Ideas regarding reverse engineering the WHOOP:

### ISSUE THREAD

**Issue by eliaspfeffer** [bWanShiTong/reverse-engineering-whoop-post#1]
Okay so what I have come up with is the following: read out the data of the sensors directly (needs physical manipulation), read out the raw data from the sensors to jump over the part where the sensor data probably gets encrypted.
Then get the raw data that got sent to whoop and the results where you need the subscription for.
Then let some datascience algorithms figure out the combination of the raw sensor data results, how to combine those together, to achieve the same results regarding the interpretation of the raw sensor data.

**Comment by bWanShiTong** [comment]
First thing is sort of problem device is very small and hard to wear if opened and tinkered, but I will try to do it.

Data doesn't get encrypted between device and app, this is actually big problem, because you are able to use device and read data from anyone near.

On data readings I am able to read heart rate and rr, for all times, I am currently unable to find out how to read temperature and blood oxygen, I think that temperature is constantly measured, but not shown, but blood oxygen is measured when device detects sleep.

**Comment by kristianschneider** [comment]
One thing I tried looking into was to disassemble the [APK ](https://apkcombo.com/whoop/com.whoop.android/download/apk) with [JDAX](https://github.com/skylot/jadx) With that you are able some degree find fun stuff. I was not able though to find a smoking gun

**Comment by bWanShiTong** [comment]
I also decompiled apk using jadx, it has few useful things, like they way app recognizes whoop device, I was able to find some code used to connect to device, but it's very slow/hard to read decompiled code, I also tried using android studios decompiler that gives about same results, but you can recompile it and run it.

On other note, I did find some useful data in databases when I pulled app directory from phone, but nothing huge

**Comment by eliaspfeffer** [comment]
@bWanShiTong maybe things are easier to decompile with Cursor these days?

**Comment by bWanShiTong** [comment]
I will try, but I am pretty sure Cursor can't decompile this

**Comment by vivek-dodia** [comment]
just came across this project. awesome work. i wanted to know if you were still actively working on this project? i really want to get a whoop band but im not a fan of subscription and so wanted to start a project like this myself and came across this. ill see what i can build off from this but thanks for all the work!

**Comment by eliaspfeffer** [comment]
I think fastest setup would be to work with cursor or windsurf to produce a
whoop clone and use the data from the whoop band. Or just use Athlytic via
Apple Watch. I went via Athlytic. No subscription, just a one time payment.
Its worth it.

Am Do., 8. Mai 2025 um 16:38 Uhr schrieb vivek ***@***.***>:

**Comment by bWanShiTong** [comment]
@vivek-dodia I am still working on it but a bit slower, for my main purpose of having heart rate monitoring 24/7 it work, now integrating other things like sleep stage detection will take more time and testing

**Comment by kristianschneider** [comment]
> [@vivek-dodia](https://github.com/vivek-dodia) I am still working on it but a bit slower, for my main purpose of having heart rate monitoring 24/7 it work, now integrating other things like sleep stage detection will take more time and testing

How have you achieved 24/7 heart rate monitoring? I am very interested in this too

**Comment by bWanShiTong** [comment]
@kristianschneider yes you can checkout code here [openwhoop](https://github.com/bWanShiTong/openwhoop)

**Comment by David456op** [comment]
@bWanShiTong  I'm wondering if there any way to get around paying for the subscription model but still being able to use the device and get data from it

**Comment by David456op** [comment]
> [@bWanShiTong](https://github.com/bWanShiTong) I'm wondering if there any way to get around paying for the subscription model but still being able to use the device and get data from it

**Comment by bWanShiTong** [comment]
Yes, check the comment above your post

**Comment by David456op** [comment]
So is that what you are currently working on, or what you have already done?

**Comment by bWanShiTong** [comment]
It is still work in progress, if you want to contribute feel free

**Comment by rawbb** [comment]
@bWanShiTong awesome work man. another potential request would be to still be able to add it to a garmin as a sensor. not sure if that's something on your radar, but i think it would open up a lot of use cases for adding it to other non subscription apps that do read and analyze some of the data. 

I also have a whoop 4 that i'd be willing to crack open and wear around collecting data open if that's something you still need. not very good at assisting with the coding portion but def willing to help with the physical device portion if needed at all!

**Comment by bWanShiTong** [comment]
Hi @rawbb I don't want to add garmin as sensor since there are already open source alternatives to read data from it, there is no need to crack open the sensor to use it with this software it should work out of the box, thanks for the enthusiasm

**Comment by rawbb** [comment]
ah, i think i misconstrued the idea. other direction. you can add _the whoop_ as a "sensor" to a garmin watch, so the data from the whoop is reviewable on the garmin app.  but as it stands you need an active whoop subscription for it to pick up that data and send it to the garmin. was wondering if using your program would allow the whoop to be picked up by the garmin watch. 

either way, sick project! i'll download it and see what i can figure out on integrating it with the garmin app.

**Comment by bWanShiTong** [comment]
That could be done, currently I am using this project to download data from whoop to my computer and do anlysis there, but in future this download will be possible on android I don't know how garmin watch does it but if integration happens from health connect or something like that it would be possible, if integration is directly between garmin and whoop I don't think it would be possible
